content/blog/2017/2017-01-31-pipeline-scalability-best-practice.adoc

```diff
@@ -0,0 +1,197 @@
+..  NonCPS functions can use a much broader subset of the Groovy
+language, such as iterators and functional features (which makes them
+more terse and fast to write).
+    ..  @NonCPS functions should not use pipeline steps internally, however
+you can store the result of a pipeline step to a variable and use it
+that as the input to a NonCPS function.
+    ... *Gotcha:* It’s not guaranteed that use of a step will generate an
+error (there is an open JIRA to implement that), but you should not rely
+on that behavior. You may see improper handling of exceptions, in
+particular.
+    ..  While normal pipeline is restricted to serializable local variables
+(see appendix at bottom), @NonCPS functions can use more complex,
+nonserializable types internally (ex regex matchers, etc). Parameters
+and return types should still be Serializable, however.
+    ... *Gotcha:* improper usages are not guaranteed to raise an error with
+normal pipeline (optimizations may mask the issue), but it is unsafe to
+rely on this behavior.
+.  *Prefer external scripts/tools for complex or CPU-expensive
+processing* rather than Groovy language features. This offloads work
+from the master to external executors, allowing for easy scale-out of
+hardware resources. It is also generally easier to test because these
+components can be tested in isolation without the full on-master
+execution environment.
+    ..  Many software vendors will provide easy command-line clients for
+their tools in various programming languages. These are often robust,
+performant, and easy to use. Plugins offer another option (see below).
+    ..  Shell or batch steps are often the easiest way to integrate these
+tools, which can be written in any language. For example: sh “java -jar
+client.jar $endPointUrl $inputData” for a Java client, or sh “python
+jiraClient.py $issueId $someParam” for a python client. +
+    ..  Examples of places to use this:
+        ...  XML/JSON parsing and transformation (ex: curl + xmllint for XML
+xpath, or curl + jq for JSON)
+        ...  Templating large files
+        ...  Nontrivial integration with external APIs (for bigger vendors,
+consider a Jenkins plugin if a quality offering exists)
+        ...  Simulations/complex calculations
+        ...  Business logic
+.  *Consider existing plugins for external integrations.* Jenkins has a
+wealth of plugins, especially for source control, artifact management,
+deployment systems, and systems automation. These can greatly reduce the
+amount of pipeline code to maintain. Well-written plugins may be be
+faster and more robust than pipeline equivalents.
+    ..  Consider both plugins and command-line clients (above) -- one may be
+easier than the other.
+    ..  Plugins may be of widely varying quality. Vendor-maintained plugins
+with a large install base are often of good quality. Third-party plugins
+with limited installations may actually be worse than writing a little
```

Comment:
"Vendor-maintained" vs. "Third-party" is a weird comparison to make. It's all basically third-party plugins. I think this statement also unfairly throws shade at smaller plugins. The advice that I give, and follow, is to look at the number of installations, the frequency of updates in the changelog, and open issues in issues.jenkins-ci.org.


Some plugins have small install counts because they have a niche functionality, it's not a question of their quality